File: app/skills/lint_markdown_clean.py
"""Lint check: forbidden internal-pipeline tokens in user-facing markdown.

Spec: ``design/sourcedata-layout.md §Naming hygiene`` and Phase 3
``§Test gate``.

Walks every ``.md`` under ``report/`` and ``future-prediction/`` for a
given date (Phase 3 scope — Phase 5 will run this corpus-wide). On each
file, greps for the forbidden-token list. Exits 0 with ``OK`` on a
clean walk, 1 with explicit per-file findings on any hit.

The lint is intentionally narrow: the goal is to prove that the new
renderer never emits Stream-jargon / parser-anchor tokens in human-
facing markdown. Phase 4 will run this against the legacy corpus once
those files have been re-rendered.

CLI:

  python -m app.skills.lint_markdown_clean --date 2026-05-04
  python -m app.skills.lint_markdown_clean --paths /tmp/synthetic.md
"""

from __future__ import annotations

import argparse
import re
import sys
from pathlib import Path


# Each entry: (token_label_for_humans, regex_pattern, flags).
# Patterns are deliberately strict — false positives here block ship.
_FORBIDDEN: list[tuple[str, str, int]] = [
    # Coined project terms.
    ("eli14", r"\beli14\b", re.IGNORECASE),
    ("JTBD", r"\bJTBD\b", 0),
    # "Stream A/B/C/D/E/F/J/K" — case-insensitive, whole-word boundary.
    ("Stream A/B/C/D/E/F/J/K", r"\bStream\s+[A-FJK]\b", 0),
    ("Reasoning trace", r"\bReasoning trace\b", re.IGNORECASE),
    # Internal IDs.
    # `Pred ID #N` (also `Prediction ID #N`).
    ("Pred ID #N", r"\bPred(?:iction)?\s*ID\s*#\d+", re.IGNORECASE),
    ("Need ID #N", r"\bNeed\s*ID\s*#\d+", re.IGNORECASE),
    ("Bridge ID #N", r"\bBridge\s*ID\s*#\d+", re.IGNORECASE),
    # Hash-style internal references like `prediction.adb89416691d7587`.
    ("prediction.<hash>", r"\bprediction\.[0-9a-f]{8,}\b", 0),
    ("need.<hash>", r"\bneed\.[0-9a-f]{8,}\b", 0),
    # Parser anchors.
    ("**Summary:** parser anchor", r"\*\*Summary\s*:\s*\*\*", 0),
    # `**Bridge (...)**` — narrowly target the parenthetical-containing
    # variant to avoid clipping legitimate prose like "## Bridge".
    ("**Bridge (...):** parser anchor", r"\*\*Bridge\s*\([^)]*\)\s*:\s*\*\*", 0),
    ("Coherence N/5", r"\bCoherence\s+\d/5\b", re.IGNORECASE),
    ("Remaining gap:", r"\bRemaining\s+gap\s*:", re.IGNORECASE),
    # Bullet keys (the literal bullet at line start).
    ("- because: bullet key", r"^\s*-\s+because\s*:", re.MULTILINE),
    ("- given: bullet key", r"^\s*-\s+given\s*:", re.MULTILINE),
    ("- so_that: bullet key", r"^\s*-\s+so_that\s*:", re.MULTILINE),
    ("- landing: bullet key", r"^\s*-\s+landing\s*:", re.MULTILINE),
    ("- eli14: bullet key", r"^\s*-\s+eli14\s*:", re.MULTILINE),
    # Pipeline meta scope prefixes — already covered by an existing
    # rule in the writer flow; restated here so a single lint pass
    # covers the full forbidden-token catalogue.
    (
        "(Tech)/(Business)/(Mix) scope prefix",
        (
            r"\((?:Tech|Non-Tech|Non-tech|Business|Biz|Mix|Technical|"
            r"Non-Technical|Technology|Tecnolog[íi]a|Tec|No-Tec|"
            r"T[ée]cnico|Negocio|Teknikal|Hindi-Teknikal|Negosyo|Halong)\)"
        ),
        0,
    ),
    # JA scope prefixes use full-width parens.
    (
        "（技術）/（ビジネス）scope prefix",
        r"（(?:技術|非技術|テクノロジー|非テクノロジー|ビジネス|非ビジネス|ビジ|ミックス)）",
        0,
    ),
]


def _scan_file(path: Path) -> list[tuple[str, int, str]]:
    """Return a list of ``(token_label, line_number, snippet)`` hits."""
    try:
        text = path.read_text(encoding="utf-8")
    except (OSError, UnicodeDecodeError) as e:
        return [(f"<unreadable: {e}>", 0, "")]
    hits: list[tuple[str, int, str]] = []
    for label, pattern, flags in _FORBIDDEN:
        for m in re.finditer(pattern, text, flags):
            line_no = text.count("\n", 0, m.start()) + 1
            # Single-line snippet, truncated.
            line_start = text.rfind("\n", 0, m.start()) + 1
            line_end = text.find("\n", m.end())
            if line_end == -1:
                line_end = len(text)
            snippet = text[line_start:line_end].strip()
            if len(snippet) > 120:
                snippet = snippet[:117] + "..."
            hits.append((label, line_no, snippet))
    return hits


def _date_paths(repo_root: Path, date_iso: str) -> list[Path]:
    """Resolve every per-locale `news-YYYYMMDD.md` + `future-prediction-…`."""
    compact = date_iso.replace("-", "")
    out: list[Path] = []
    for locale in ("en", "ja", "es", "fil"):
        out.append(repo_root / "report" / locale / f"news-{compact}.md")
        out.append(
            repo_root
            / "future-prediction"
            / locale
            / f"future-prediction-{compact}.md"
        )
    return [p for p in out if p.is_file()]


def _safe_print(line: str) -> None:
    """Print a line that may contain Unicode the console can't encode.

    On Windows, the default cp1252 console raises ``UnicodeEncodeError``
    on any character outside Latin-1. Lint output frequently includes
    snippets with ``≥``, ``→``, etc., which would crash the runner mid-
    iteration. Re-encode through the stdout encoding with replacement
    so the lint can finish enumerating hits.
    """
    enc = (sys.stdout.encoding or "utf-8").lower()
    try:
        print(line)
    except UnicodeEncodeError:
        print(line.encode(enc, errors="replace").decode(enc, errors="replace"))


def lint_paths(paths: list[Path]) -> int:
    """Lint a concrete path list. Returns the desired exit code."""
    if not paths:
        _safe_print("OK lint-markdown-clean: no files to check")
        return 0
    failed = False
    for path in paths:
        hits = _scan_file(path)
        if hits:
            failed = True
            _safe_print(f"FAIL {path}: {len(hits)} forbidden-token hit(s)")
            for label, line_no, snippet in hits:
                _safe_print(f"  - line {line_no}: {label}")
                if snippet:
                    _safe_print(f"    {snippet}")
        else:
            _safe_print(f"OK {path}")
    if failed:
        return 1
    _safe_print(f"OK lint-markdown-clean: {len(paths)} file(s) clean")
    return 0


def main(argv: list[str] | None = None) -> int:
    p = argparse.ArgumentParser(
        description=(
            "Greps user-facing markdown for forbidden internal-pipeline "
            "tokens (Stream A-K, Pred ID #N, parser anchors, eli14, ...)."
        )
    )
    g = p.add_mutually_exclusive_group(required=True)
    g.add_argument(
        "--date",
        help=(
            "ISO date YYYY-MM-DD. Lints every per-locale news-YYYYMMDD.md "
            "and future-prediction-YYYYMMDD.md under the repo."
        ),
    )
    g.add_argument(
        "--paths",
        nargs="+",
        help="Explicit path list (overrides --date).",
    )
    p.add_argument(
        "--repo-root",
        default=".",
        help="Repository root (defaults to cwd).",
    )
    args = p.parse_args(argv)

    repo_root = Path(args.repo_root).resolve()
    if args.paths:
        paths = [Path(p_) for p_ in args.paths]
    else:
        paths = _date_paths(repo_root, args.date)
    return lint_paths(paths)


if __name__ == "__main__":  # pragma: no cover
    sys.exit(main())

File: engines/nunc-fluens/pipeline/src/render/lint-markdown-clean.ts
// TS port of app/skills/lint_markdown_clean.py — forbidden internal-
// pipeline tokens in user-facing markdown (design/sourcedata-layout.md
// §Naming hygiene + the ADR-002 anti-inertia vocabulary).
import { existsSync, readFileSync } from 'node:fs';
import { join } from 'node:path';
import { DAILY_NEWS_REL, FP_REL, LOCALES } from '../world-paths.ts';

// Each entry: [human label, regex]. The 'g' flag is added at scan time;
// 'i'/'m' mirror the oracle's re.IGNORECASE / re.MULTILINE.
const FORBIDDEN: Array<[string, RegExp]> = [
  ['plain_language', /\bplain_language\b/i],
  ['JTBD', /\bJTBD\b/],
  ['legacy stream-letter label', /\bStream\s+[A-FJK]\b/],
  ['Reasoning trace', /\bReasoning trace\b/i],
  ['Pred ID #N', /\bPred(?:iction)?\s*ID\s*#\d+/i],
  ['Need ID #N', /\bNeed\s*ID\s*#\d+/i],
  ['Bridge ID #N', /\bBridge\s*ID\s*#\d+/i],
  ['prediction.<hash>', /\bprediction\.[0-9a-f]{8,}\b/],
  ['need.<hash>', /\bneed\.[0-9a-f]{8,}\b/],
  ['**Summary:** parser anchor', /\*\*Summary\s*:\s*\*\*/],
  ['**Bridge (...):** parser anchor', /\*\*Bridge\s*\([^)]*\)\s*:\s*\*\*/],
  ['Coherence N/5', /\bCoherence\s+\d\/5\b/i],
  ['Remaining gap:', /\bRemaining\s+gap\s*:/i],
  ['- because: bullet key', /^\s*-\s+because\s*:/m],
  ['- given: bullet key', /^\s*-\s+given\s*:/m],
  ['- so_that: bullet key', /^\s*-\s+so_that\s*:/m],
  ['- landing: bullet key', /^\s*-\s+landing\s*:/m],
  ['- plain_language: bullet key', /^\s*-\s+plain_language\s*:/m],
  // Detector, not stripper: the canonical strip list lives in
  // src/export/prefix-tokens.json (P8); these stay an independent subset.
  ['(Tech)/(Business)/(Mix) scope prefix',
    /\((?:Tech|Non-Tech|Non-tech|Business|Biz|Mix|Technical|Non-Technical|Technology|Tecnolog[íi]a|Tec|No-Tec|T[ée]cnico|Negocio|Teknikal|Hindi-Teknikal|Negosyo|Halong)\)/],
  ['（技術）/（ビジネス）scope prefix',
    /（(?:技術|非技術|テクノロジー|非テクノロジー|ビジネス|非ビジネス|ビジ|ミックス)）/],
  ['day-N storyline numbering',
    /\bday-(?:\d+|twenty|thirty|forty|fifty|sixty|seventy|eighty|ninety|one|two|three|four|five|six|seven|eight|nine|ten|eleven|twelve|thirteen|fourteen|fifteen|sixteen|seventeen|eighteen|nineteen)(?:-(?:one|two|three|four|five|six|seven|eight|nine))?\b/],
  ['aging vocabulary',
    /\b(?:weekend|doubly|triply|quadruply|quintuply|sextuply)[\s-](?:weekend[\s-])?aged\b/i],
  ['N-day-old artifact filler',
    /\b(?:\d+|one|two|three|four|five|six|seven|eight|nine|ten)[\s-]day[\s-]old\s+artifact\b/i],
  ['holiday-equivalent day filler',
    /\bholiday[\s-]equivalent\s+(?:day|operating\s+day|days)\b/i],
];

export interface LintHit { label: string; line: number; snippet: string }

export function scanText(text: string): LintHit[] {
  const hits: LintHit[] = [];
  for (const [label, pattern] of FORBIDDEN) {
    const re = new RegExp(pattern.source, pattern.flags + 'g');
    for (const m of text.matchAll(re)) {
      const start = m.index ?? 0;
      const line = (text.slice(0, start).match(/\n/g) ?? []).length + 1;
      const lineStart = text.lastIndexOf('\n', start - 1) + 1;
      let lineEnd = text.indexOf('\n', start + m[0].length);
      if (lineEnd === -1) lineEnd = text.length;
      let snippet = text.slice(lineStart, lineEnd).trim();
      if (snippet.length > 120) snippet = snippet.slice(0, 117) + '...';
      hits.push({ label, line, snippet });
    }
  }
  return hits;
}

/** The day's user-facing markdown files, over the FULL render set
 * ('en' + the effective non-EN set; default = the universe). */
export function datePaths(
  publishRoot: string, dateIso: string, locales: readonly string[] = LOCALES,
): string[] {
  const compact = dateIso.replaceAll('-', '');
  const out: string[] = [];
  for (const locale of locales) {
    out.push(join(publishRoot, DAILY_NEWS_REL, locale, `news-${compact}.md`));
    out.push(join(publishRoot, FP_REL, locale, `future-prediction-${compact}.md`));
  }
  return out.filter(p => existsSync(p));
}

export interface LintResult { exit: number; lines: string[] }

export function lintPaths(paths: string[]): LintResult {
  const lines: string[] = [];
  if (paths.length === 0) {
    lines.push('OK lint-markdown-clean: no files to check');
    return { exit: 0, lines };
  }
  let failed = false;
  for (const path of paths) {
    const hits = scanText(readFileSync(path, 'utf8'));
    if (hits.length) {
      failed = true;
      lines.push(`FAIL ${path}: ${hits.length} forbidden-token hit(s)`);
      for (const h of hits) {
        lines.push(`  - line ${h.line}: ${h.label}`);
        if (h.snippet) lines.push(`    ${h.snippet}`);
      }
    } else {
      lines.push(`OK ${path}`);
    }
  }
  if (failed) return { exit: 1, lines };
  lines.push(`OK lint-markdown-clean: ${paths.length} file(s) clean`);
  return { exit: 0, lines };
}
